File: DataStructure/JavaScript/StackImp/Stack.js
'use strict';

const Node = require('./Node');

class Stack {
  constructor() {
    this.top = null;
  }

  push(data) {
    const node = new Node(data);
    if (!this.top) {
      this.top = node;
      return this;
    }
    node.next = this.top;
    this.top = node;
    return this;
  }

  pop() {
    const removed = this.top.data;
    this.top = this.top.next;
    return removed;
  }

  peek() {
    return this.top.data;
  }

  display() {
    let output = '';
    let current = this.top;
    while (current) {
      output += `${current.data} --> `;
      current = current.next;
    }
    return output;
  }

  isEmpty() {
    return this.top ? false : true;
  }

  length() {
    let counter = 0;
    let current = this.top;
    while (current) {
      counter++;
      current = current.next;
    }
    return counter;
  }
}

module.exports = Stack;

File: DataStructure/Java/SatckImp/StackImpLL/Stack.java
package SatckImp.StackImpLL;

public class Stack<T> {

    private Node<T> top;

    void push(T data){
        Node<T> newNode = new Node<>(data);
        newNode.next = top;
        top = newNode;
    }

    String display(){
        StringBuilder sb = new StringBuilder();
        Node<T> current = top;
        while (current != null){
            sb.append(current.data).append(" --> ");
            current = current.next;
        }
        sb.append("NULL");
        return sb.toString();
    }

    T peek(){
        return top.data;
    }

    T pop(){
        T wanted = top.data;
        top = top.next;
        return wanted;
    }

    boolean isEmpty(){
        return top == null;
    }
}
